From 0062d98045a375e0676a2c9e4a1ddbb17c43294d Mon Sep 17 00:00:00 2001
From: Dmitry Ivanov <idmitry@qti.qualcomm.com>
Date: Wed, 12 May 2021 08:12:50 +0200
Subject: [PATCH] vrend: Disable dual_src_blend feature
 (GL_EXT_blend_func_extended) for Qualcomm.

The dual_src_blend feature as in support for GL_EXT_blend_func_extended
seems to be broken on Qualcomm Adreno. Reports from the past about broken
MRT (multi render target) support can be found on the internet as well,
so it looks like that didn't get fixed, yet.

The error messages happen at the linking stage, and follow a pattern similar
to this: "Error: Output ... location or component exceeds max allowed."
where the output variable is part of the multiple outputs from the fragment
shader stage.

With this change the feature is suppressed even if the driver reports
GL_EXT_blend_func_extended being available. Being an optional feature,
virglrenderer will adapt accordingly. This fix can also be used on
other platforms, however, it may limit their capabilities this way.
Exynos/Mali doesn't support this extension, for example, so it is safe
to use there.

Note: This is a workaround for a OpenGL/ES driver issue.

Signed-off-by: Dmitry Ivanov <idmitry@qti.qualcomm.com>
---
 src/vrend_renderer.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/vrend_renderer.c b/src/vrend_renderer.c
index 0a323c3a..87c59e33 100644
--- a/src/vrend_renderer.c
+++ b/src/vrend_renderer.c
@@ -422,7 +422,7 @@ static inline bool has_feature(enum features_id feature_id)
    VREND_DEBUG(dbg_feature_use, NULL, "Try using feature %s:%d\n",
                feature_list[feature_id].log_name,
                retval);
-   return retval;
+   return (feature_id != feat_dual_src_blend) && retval;
 }
 
 
-- 
2.43.2

